server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -562,6 +562,10 @@ private static void parseArray(ParseContext context, ObjectMapper parentMapper,
         }
     }
 
+    private static boolean parsesArrayValue(Mapper mapper) {
+        return mapper instanceof FieldMapper && ((FieldMapper) mapper).parsesArrayValue();
```

Comment:
Unfortunately this refactor didn't actually cut down on the number of `instanceof` checks. But I feel better about this one, because it's very common for us to check `instanceof` against `FieldMapper` or `ObjectMapper` -- it's just part of how mappings + document parsing currently work.